cellarium/ml/models/cellarium_gpt.py

```diff
@@ -0,0 +1,745 @@
+        n_context: int = 2048,
+        attn_mult: float = 1.0,
+        input_mult: float = 1.0,
+        output_mult: float = 1.0,
+        initializer_range: float = 0.02,
+        backend: Literal["keops", "torch", "math", "flash", "mem_efficient"] = "keops",
+        log_metrics: bool = True,
+    ) -> None:
+        super().__init__()
+        self.var_names_g = np.array(var_names_g)
+        self.n_vars = len(var_names_g)
+        self.gpt_model = GPTModel(
+            self.n_vars,
+            d_model,
+            d_ffn,
+            n_heads,
+            n_blocks,
+            dropout,
+            use_bias,
+            attn_mult,
+            input_mult,
+            backend,
+        )
+        self.nb_head = NegativeBinomialHead(d_model, use_bias, output_mult)
+        assert n_context <= self.n_vars + 1, "n_context must be less than or equal to the number of genes + 1"
+        self.n_context = n_context
+        self.initializer_range = initializer_range
+        self.reset_parameters()
+
+        self.log_metrics = log_metrics
+
+    def reset_parameters(self) -> None:
+        self.apply(self._init_weights)
+
+    def _init_weights(self, module) -> None:
+        """Initialize the weights."""
+        if isinstance(module, nn.Linear):
+            # Slightly different from the TF version which uses truncated_normal for initialization
+            # cf https://github.com/pytorch/pytorch/pull/5617
+            module.weight.data.normal_(mean=0.0, std=self.initializer_range)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, nn.Embedding):
+            # assert self.optimizer == "adam", "Only Adam(W) optimizer is supported for now."
+            module.weight.data.normal_(mean=0.0, std=self.initializer_range)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+        elif isinstance(module, nn.LayerNorm):
+            if module.bias is not None:
+                module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+
+        # Reinitialize selected weights subject to the OpenAI GPT-2 Paper Scheme:
+        #   > A modified initialization which accounts for the accumulation on the residual path with model depth. Scale
+        #   > the weights of residual layers at initialization by a factor of 1/√N where N is the # of residual layers.
+        #   >   -- GPT-2 :: https://openai.com/blog/better-language-models/
+        #
+        # Reference (Megatron-LM): https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py
+        for name, p in module.named_parameters():
+            if name == "dense2.weight" or name == "Wo.weight":
+                # Special Scaled Initialization --> There are 2 Layer Norms per Transformer Block
+                p.data.normal_(mean=0.0, std=(self.initializer_range / math.sqrt(2 * self.gpt_model.n_blocks)))
+
+    def tokenize(
```

Comment:
documentation

Reply:
Can you also separately take `obs_total_mrna_umis_n` (to be concatenated to `values_nc`) and `out_total_mrna_umis_n` to generate a token specifying the library size we need the _readout_ to be at? This construct can be used for model training with downsampling. We could use Brice's logic to `Duplicate` the counts, downsample the first copy to use for observation, and use the second copy for readout.

Another thought is to have multiple `tokenize` methods. For example:
-- `generate_observation_tokens`
-- `generate_output_tokens`
-- `generate_register_tokens` (later)
-- `generate_metadata_tokens` (later)

In this construct, `generate_observation_tokens` takes `x_ng`, `obs_total_mrna_umis_n` as input and generates a bunch of tokens.  `generate_output_tokens`, for now, just takes `out_total_mrna_umis_n` and generates a single token.

All of the tokens are then concatenated and given to an embedding layer. The tokens should also be equipped with metadata to specify how they should be embedded ...

Reply:
Make an issue (#131).